From 572b63960f59f1da22aa815ce93b36c1c4a3a4c6 Mon Sep 17 00:00:00 2001
From: Benjamin Cheng <benjamin.cheng@amd.com>
Date: Wed, 12 Nov 2025 13:49:57 -0500
Subject: [PATCH] radv/video: Align each layer of encode DPB to 256

VCN requires the luma/chroma VAs to be 256 aligned. On VCN5, the
collocated buffer was not 256 aligned which can cause these VAs to be
unaligned.

This fixes VVL PositiveVideoEncodeH264.Basic on VCN5.

Fixes: 37e71a5cb24 ("radv/video: add support for AV1 encoding")
Reviewed-by: David Rosca <david.rosca@amd.com>
Part-of: <https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/38408>
(cherry picked from commit 884849587569919b7ef40021da70e6858d93cbe9)
---
 .pick_status.json               |  2 +-
 src/amd/vulkan/radv_video_enc.c | 16 ++++++++++------
 2 files changed, 11 insertions(+), 7 deletions(-)

diff --git a/.pick_status.json b/.pick_status.json
index 799a5e1424106..663b01e32a916 100644
--- a/.pick_status.json
+++ b/.pick_status.json
@@ -1094,7 +1094,7 @@
         "description": "radv/video: Align each layer of encode DPB to 256",
         "nominated": true,
         "nomination_type": 2,
-        "resolution": 0,
+        "resolution": 1,
         "main_sha": null,
         "because_sha": "37e71a5cb24cd6053603ca64b67352e0c8e2fce8",
         "notes": null
diff --git a/src/amd/vulkan/radv_video_enc.c b/src/amd/vulkan/radv_video_enc.c
index 2a1098d0b38dd..c10ca787c5520 100644
--- a/src/amd/vulkan/radv_video_enc.c
+++ b/src/amd/vulkan/radv_video_enc.c
@@ -1543,6 +1543,7 @@ radv_enc_ctx2(struct radv_cmd_buffer *cmd_buffer, const VkVideoEncodeInfoKHR *in
          metadata_size += RENCODE_AV1_FRAME_CONTEXT_CDF_TABLE_SIZE;
          metadata_size += RENCODE_AV1_CDEF_ALGORITHM_FRAME_CONTEXT_SIZE;
       }
+      metadata_size = align(metadata_size, ENC_ALIGNMENT);
 
       uint32_t dpb_array_idx = res->baseArrayLayer + dpb_iv->vk.base_array_layer;
       uint64_t luma_va = dpb_img->bindings[0].addr + dpb_array_idx * (luma_size + chroma_size + metadata_size);
@@ -3401,17 +3402,20 @@ radv_video_get_enc_dpb_image(struct radv_device *device, const struct VkVideoPro
    }
 
    for (unsigned i = 0; i < num_reconstructed_pictures; i++) {
-      image->size += luma_size;
-      image->size += chroma_size;
+      unsigned metadata_size = 0;
       if (is_av1) {
-         image->size += RENCODE_AV1_FRAME_CONTEXT_CDF_TABLE_SIZE;
-         image->size += RENCODE_AV1_CDEF_ALGORITHM_FRAME_CONTEXT_SIZE;
+         metadata_size += RENCODE_AV1_FRAME_CONTEXT_CDF_TABLE_SIZE;
+         metadata_size += RENCODE_AV1_CDEF_ALGORITHM_FRAME_CONTEXT_SIZE;
       }
       if (pdev->enc_hw_ver >= RADV_VIDEO_ENC_HW_5) {
-         image->size += RENCODE_MAX_METADATA_BUFFER_SIZE_PER_FRAME;
+         metadata_size += RENCODE_MAX_METADATA_BUFFER_SIZE_PER_FRAME;
          if (has_h264_b_support)
-            image->size += colloc_bytes;
+            metadata_size += colloc_bytes;
       }
+
+      image->size += luma_size;
+      image->size += chroma_size;
+      image->size += align(metadata_size, ENC_ALIGNMENT);
    }
    image->alignment = ENC_ALIGNMENT;
 }
-- 
GitLab

